### PR TITLE
[SVE][CodeGenPrepare] Sink address calculations that match SVE gather/scatter addressing modes.

### PR DESCRIPTION
**Pull request by paulwalker-arm**
SVE supports scalar+vector and scalar+extw(vector) addressing modes.
However, the masked gather/scatter intrinsics take a vector of
addresses, which means address computations can be hoisted out of
loops.  The is especially true for things like offsets where the
true size of offsets is lost by the time you get to code generation.

This is problematic because it forces the code generator to legalise
towards `<vscale x 2 x ty>` vectors that will not maximise bandwidth
if the main block datatypes is in fact i32 or smaller.

This patch sinks GEPs and extends for cases where one of the above
addressing modes can be used.

NOTE: There are cases where it would be better to split the extend
in two with one half hoisted out of a loop and the other within the
loop.  Whilst true I think this switch of default is still better
than before because the extra extends are an improvement over being
forced to split a gather/scatter.

### CODE REVIEW

**Review by sdesmalen-arm — COMMENTED** [review]
Looks sensible to me, just left one question.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 1]
```diff
@@ -14380,6 +14380,31 @@ static bool areOperandsOfVmullHighP64(Value *Op1, Value *Op2) {
...
+  Value *Offsets = GEP->getOperand(1);
+
+  // We only care about scalar_base+vector_offsets.
+  if (Base->getType()->isVectorTy() || !Offsets->getType()->isVectorTy())
+    return false;
+
+  // Sink extends that would allow us to use 32-bit offset vectors.
+  if (isa<SExtInst>(Offsets) || isa<ZExtInst>(Offsets)) {
+    auto *OffsetsInst = cast<Instruction>(Offsets);
+    if (OffsetsInst->getType()->getScalarSizeInBits() > 32 &&
```
[the remark this answers is not in the document]
Comment (paulwalker-arm) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:14399]: I just wanted to maximise coverage. For example, offsets like `sext <vscale x 2 x i32> %0 to <vscale x 2 x i48>` can also be handled more efficiently when the code generator can see the extend.

[the remark this answers is not in the document]
Comment (sdesmalen-arm) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:14399]: Makes sense, that would be type legalised to `<vscale x 2 x i64>` and you'd still get the same benefit.

**Review by sdesmalen-arm — APPROVED** [review]
LGTM

**Review by huntergr-arm — APPROVED** [review]
LGTM

**Pull request merged**